### PR TITLE
Center window

### LINKED ISSUE #20012: Center rolling window with date NotImplementedError
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

idx = pd.date_range(start="2010-01-01", end="2010-01-30", fred="1D")
df = pd.DataFrame(index=idx, data=np.sin(np.linspace(0,2*np.pi,len(idx))))

# works well
df.rolling(5, center=True, min_periods=5)

# fail with 
# NotImplementedError: center is not implemented for datetimelike and offset based windows
df.rolling("5D", center=True, min_periods=5)
```

Complete traceback:
<details>

```python

NotImplementedError                       Traceback (most recent call last)
<ipython-input-54-058f86b11afc> in <module>()
----> 1 df.rolling("5D", center=True, min_periods=5)

~/.local/lib/python3.5/site-packages/pandas/core/generic.py in rolling(self, window, min_periods, freq, center, win_type, on, axis, closed)
   7063                                    min_periods=min_periods, freq=freq,
   7064                                    center=center, win_type=win_type,
-> 7065                                    on=on, axis=axis, closed=closed)
   7066 
   7067         cls.rolling = rolling

~/.local/lib/python3.5/site-packages/pandas/core/window.py in rolling(obj, win_type, **kwds)
   2064         return Window(obj, win_type=win_type, **kwds)
   2065 
-> 2066     return Rolling(obj, **kwds)
   2067 
   2068 

~/.local/lib/python3.5/site-packages/pandas/core/window.py in __init__(self, obj, window, min_periods, freq, center, win_type, axis, on, closed, **kwargs)
     84         self.win_freq = None
     85         self.axis = obj._get_axis_number(axis) if axis is not None else None
---> 86         self.validate()
     87 
     88     @property

~/.local/lib/python3.5/site-packages/pandas/core/window.py in validate(self)
   1104             # we don't allow center
   1105             if self.center:
-> 1106                 raise NotImplementedError("center is not implemented "
   1107                                           "for datetimelike and offset "
   1108                                           "based windows")

NotImplementedError: center is not implemented for datetimelike and offset based windows

```

</details>

#### Problem description

The rolling window is not able to be centered with a datetimelike index, which is a problem when dealing with time series. 

I think that this is a WIP feature. I don't see anything blocking about it.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-3-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: C.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: 3.0.6
pip: 9.0.1
setuptools: 38.2.3
Cython: None
numpy: 1.13.3
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: 3.3.0
numexpr: 2.6.1
feather: None
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: 3.7.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.2.2
pymysql: None
psycopg2: 2.6.2 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by adamamiller**
- [x] closes #20012
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Added center functionality for `VariableWindowIndexer`. Note - I am unsure if the NotImplementedError in lines 1966-1969 in rolling.py still correctly raises an error for offset based windows.

### CODE REVIEW

File: pandas/_libs/window/indexers.pyx [hunk 1]
```diff
@@ -76,14 +79,27 @@ def calculate_variable_window_bounds(
     # right endpoint is open
     else:
         end[0] = 0
+    if center_window:
+        for j in range(0, num_values+1):
+            if (index[j] == index[0] + index_growth_sign*window_size/2 and
```
Comment (jreback) [inline pandas/_libs/window/indexers.pyx:84]: this is likley failing linting (need spaces between operators)

**Review by jreback — CHANGES_REQUESTED** [review]
needs a fair amount of tests

**Review by arw2019 — COMMENTED** [review]
@adamamiller are you interested in continuing? If yes please address @jreback comment (he's requesting tests) and merge master/resolve conflicts

**Comment by arw2019** [comment]
Closing for now. @adamamiller ping whenever you'd like to continue and we'll reopen!

**Comment by Zaubeerer** [comment]
Would love to see this functionality and would like to pick up the pull request with @luholgit

What are the next required steps?

We already forked the pandas repo and checked out the branch from @adamamiller.

However, when trying to update the branch by pulling from origin/master into fork/center_window we encountered several merge conflicts which we are not sure how to solve.

Should we solve them based on our understanding and then discuss those changes in this PR?

**Comment by Zaubeerer** [comment]
@jreback and @arw2019, mentioning you explicitly as we need this change urgently and would love to implement at the open source directly as opposed to our private repository! :)

**Comment by jreback** [comment]
you see the comments above

this needs to have master merged and needs tests

**Comment by Zaubeerer** [comment]
Thanks for your quick response, we are implementing the merge and tests right now.

(Un)Fortunately, we were not able to push to @adamamiller's fork. 

Could you change the pull request to come from https://github.com/luholgit/pandas/tree/center_window?

**Comment by Zaubeerer** [comment]
Or shall we create a new issue such that the new source branch from our fork is selected automatically?

**Comment by MarcoGorelli** [comment]
@Zaubeerer thanks for taking this forwards - you can open a new pull request

**Comment by jreback** [comment]
@Zaubeerer you just create a new PR, no need to create a new issue

**Pull request closed without merging**